File: src/leetcode/m289/GameOfLife.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m289;

import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode 289
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/09/27 09:11
 */
public class GameOfLife {
    private final int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1},
            {-1, -1}, {-1, 1}, {1, -1}, {1, 1}};

    public void gameOfLife(int[][] board) {
        List<int[]> lifeChanges = new ArrayList<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                int count = countLiveNeighbors(i, j, board);
                if (count < 2 && board[i][j] == 1) {
                    lifeChanges.add(new int[]{i, j});
                } else if (count > 3 && board[i][j] == 1) {
                    lifeChanges.add(new int[]{i, j});
                } else if (count == 3 && board[i][j] == 0) {
                    lifeChanges.add(new int[]{i, j});
                }
            }
        }

        for (int[] change : lifeChanges) {
            int i = change[0];
            int j = change[1];
            board[i][j] = 1 - board[i][j];
        }
    }

    private int countLiveNeighbors(int i, int j, int[][] board) {
        int count = 0;
        for (int[] direction : directions) {
            int m = i + direction[0];
            int n = j + direction[1];
            if (m >= 0 && m < board.length && n >= 0 && n < board[0].length && board[m][n] == 1) {
                count++;
            }
        }
        return count;
    }
}

File: src/leetcode/m289/GameOfLife.go
package m289

// LeetCode 289
func gameOfLife(board [][]int) {
	switched := make([][]int, 0)
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			coords := []int{i, j}
			neighbors := countNeighbors(board, coords)
			if board[i][j] == 1 {
				if neighbors < 2 || neighbors > 3 {
					switched = append(switched, coords)
				}
			} else if neighbors == 3 {
				switched = append(switched, coords)
			}
		}
	}

	for _, sw := range switched {
		board[sw[0]][sw[1]] ^= 1
	}
}

func countNeighbors(board [][]int, coords []int) int {
	count := 0
	directions := [][]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}, {1, 1}, {-1, -1}, {-1, 1}, {1, -1}}
	for _, dir := range directions {
		x := coords[0] + dir[0]
		y := coords[1] + dir[1]
		if x >= 0 && x < len(board) && y >= 0 && y < len(board[0]) && board[x][y] == 1 {
			count++
		}
	}
	return count
}
